presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,578 @@
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stRelate(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right, @SqlType(StandardTypes.VARCHAR) Slice relation)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        verifySameSpatialReference(leftGeometry, rightGeometry);
+        return leftGeometry.relate(rightGeometry, relation.toStringUtf8());
+    }
+
+    @SqlNullable
+    @Description("Returns TRUE if the geometries have at least one point in common, but their interiors do not intersect")
+    @ScalarFunction("ST_Touches")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stTouches(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        verifySameSpatialReference(leftGeometry, rightGeometry);
+        return leftGeometry.touches(rightGeometry);
+    }
+
+    @SqlNullable
+    @Description("Returns TRUE if the geometry A is completely inside geometry B")
+    @ScalarFunction("ST_Within")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stWithin(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        verifySameSpatialReference(leftGeometry, rightGeometry);
+        return leftGeometry.within(rightGeometry);
+    }
+
+    private static OGCGeometry getGeometry(OGCGeometry geometry, String function, String...validTypes)
+    {
+        requireNonNull(geometry, "geometry is null");
+        String type = geometry.geometryType();
+        for (String validType : validTypes) {
+            if (type.equals(validType)) {
+                return geometry;
+            }
+        }
+        throw new IllegalArgumentException(function + " only applies to " + StringUtils.join(asList(validTypes), " or ") + ". Input type is: " + type);
+    }
+
+    private static Envelope getEnvelope(OGCGeometry geometry)
+    {
+        Envelope envelope = new Envelope();
+        geometry.getEsriGeometry().queryEnvelope(envelope);
+        return envelope;
+    }
+
+    private static void verifySameSpatialReference(OGCGeometry leftGeometry, OGCGeometry rightGeometry)
+    {
+        checkArgument(leftGeometry.getEsriSpatialReference().equals(rightGeometry.getEsriSpatialReference()), "Input Geometries Spatial Reference do not match");
+    }
+
+    // centroid is arithmetic mean of the input points
+    private static Point getPointsCentroid(OGCGeometry geometry)
+    {
+        String type = geometry.geometryType();
+        int pointCount = ((MultiVertexGeometry) geometry.getEsriGeometry()).getPointCount();
+        double xSum = 0;
+        double ySum = 0;
+        for (int i = 0; i < pointCount; i++) {
+            Point point = null;
+            if (type.equals(MULTI_POINT)) {
```

Comment:
This logic can be simplified by capturing the result of (MultiVertexGeometry) geometry.getEsriGeometry() in a variable in the beginning of the method and using MultiVertexGeometry.getPoint(i) method.